homeassistant/components/binary_sensor/rflink.py

```diff
@@ -0,0 +1,114 @@
+"""
+Support for Rflink binary sensors.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.rflink/
+"""
+import logging
+
+from homeassistant.components.binary_sensor import (
+    DEVICE_CLASSES_SCHEMA)
+from homeassistant.components.rflink import (
+    CONF_ALIASES, CONF_DEVICES, DATA_ENTITY_LOOKUP, DOMAIN,
+    EVENT_KEY_COMMAND, RflinkDevice, cv, vol)
+from homeassistant.const import (
+    CONF_FORCE_UPDATE, CONF_NAME, CONF_DEVICE_CLASS, CONF_PLATFORM, STATE_OFF,
+    STATE_ON)
+import homeassistant.helpers.event as evt
+
+CONF_OFF_DELAY = 'off_delay'
+DEFAULT_FORCE_UPDATE = False
+
+DEPENDENCIES = ['rflink']
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORM_SCHEMA = vol.Schema({
+    vol.Required(CONF_PLATFORM): DOMAIN,
+    vol.Optional(CONF_DEVICES, default={}): vol.Schema({
+        cv.string: {
+            vol.Optional(CONF_NAME): cv.string,
+            vol.Optional(CONF_DEVICE_CLASS):
+                DEVICE_CLASSES_SCHEMA,
+            vol.Optional(CONF_FORCE_UPDATE, default=DEFAULT_FORCE_UPDATE):
+                cv.boolean,
+            vol.Optional(CONF_OFF_DELAY):
+                vol.All(vol.Coerce(int), vol.Range(min=0)),
+            vol.Optional(CONF_ALIASES, default=[]):
+                vol.All(cv.ensure_list, [cv.string]),
+        },
+    }),
+})
+
+
+def devices_from_config(domain_config, hass=None):
+    """Parse configuration and add Rflink sensor devices."""
+    devices = []
+    for device_id, config in domain_config[CONF_DEVICES].items():
+        device = RflinkBinarySensor(device_id, hass, **config)
+        devices.append(device)
+
+        # Register entity (and aliases) to listen to incoming rflink events
+        hass.data[DATA_ENTITY_LOOKUP][
+            EVENT_KEY_COMMAND][device_id].append(device)
```

Comment:
Please refactor this. We shouldn't store entities in a shared dictionary in hass.data. Entities should only be accessed directly from within their respective platforms.

Use our dispatch helper instead to signal the entity methods from the component.

See the tuya component for example.

Reply:
Can you explain in a little more detail what you mean, maybe point to relevant function in the tuya component?

Reply:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/tuya.py#L100-L102

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/tuya.py#L129-L130

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/tuya.py#L100-L102

Reply:
Thanks!

Reply:
@MartinHjelmare rfxtrx is functionality wise closer to rflink than tuya.
Can you comment on if the pattern used in rfxtrx is OK, especially `RECEIVED_EVT_SUBSCRIBERS` and `RFX_DEVICES` in components\rfxtrx.py

Reply:
No the pattern in rfxtrx is not ok. It also stores entities in a dictionary in the wrong place, ie outside their platforms. It also uses constants for things that are not constant. Bad!

Please use the our dispatch helper to communicate between component and platforms.

I agree this is a big refactor and it should be done separately. We want to merge the refactor first then rebase this PR and merge it after the refactor is done.

Reply:
@emontnemery If you want I can help you with the tests. I can test the RFLink cover, light, sensor and switch physical devices.
I can also try to help with the refactor (only devices implementation, not the base ones), but I must admit that I'm not very familiar with python (although I'm learning on the fly)

Reply:
@MartinHjelmare OK, I got it. I hope to have time to work on this and open a PR for review next week. Would you mind explaining _why_ it's not OK to store entities in a dictionary as is done in both RFLink and rfxtrx though?
@javicalle Thanks, I'll ping you once the refactoring is ready for test.

Reply:
Storing entities in dictionaries is not bad per se. It's when we do that in a shared dictionary to be able to access the entity directly from outside its platform it becomes bad. We want to have a decoupled architecture where entity methods are a detail of their implementation. Communication to and from entities should either go via core eg service calls and state updates, or via our dispatch helper that allows interaction without knowing the entity implementation.